mmaction/datasets/ssn_dataset.py

```diff
@@ -0,0 +1,815 @@
+            }
+        self.logger.info('NMS finished')
+
+        # get gts
+        all_gts = self.get_all_gts()
+        for class_idx in range(len(detections)):
+            if (class_idx not in all_gts):
+                all_gts[class_idx] = dict()
+
+        # get predictions
+        plain_detections = {}
+        for class_idx in range(len(detections)):
+            detection_list = []
+            for vid, dets in detections[class_idx].items():
+                detection_list.extend([[vid, class_idx] + x[:3]
+                                       for x in dets.tolist()])
+            plain_detections[class_idx] = detection_list
+
+        eval_results = {}
+        for metric in metrics:
+            if metric == 'mAP':
+                if eval_dataset == 'thumos14':
+                    iou_range = np.arange(0.1, 1.0, .1)
+                    ap_values = eval_ap(plain_detections, all_gts, iou_range)
+                    map_ious = ap_values.mean(axis=0)
+                    self.logger.info('Evaluation finished')
+
+                    for iou, map_iou in zip(iou_range, map_ious):
+                        eval_results[f'mAP@{iou:.02f}'] = map_iou
+
+        return eval_results
+
+    def construct_proposal_pools(self):
+        """Construct positve proposal pool, incomplete proposal pool and
+        background proposal pool of the entire dataset."""
+        for video_info in self.video_infos:
+            positives = self.get_positives(
+                video_info['gts'], video_info['proposals'],
+                self.assigner.positive_iou_threshold,
+                self.sampler.add_gt_as_proposals)
+            self.positive_pool.extend([(video_info['video_id'], proposal)
+                                       for proposal in positives])
+
+            incompletes, backgrounds = self.get_negatives(
+                video_info['proposals'],
+                self.assigner.incomplete_iou_threshold,
+                self.assigner.background_iou_threshold,
+                self.assigner.background_coverage_threshold,
+                self.assigner.incomplete_overlap_threshold)
+            self.incomplete_pool.extend([(video_info['video_id'], proposal)
+                                         for proposal in incompletes])
+            self.background_pool.extend([video_info['video_id'], proposal]
+                                        for proposal in backgrounds)
+
+    def get_all_gts(self):
+        """Fetch groundtruth instances of the entire dataset."""
+        gts = {}
+        for video_info in self.video_infos:
+            vid = video_info['video_id']
+            for gt in video_info['gts']:
+                class_idx = gt.label - 1
+                # gt_info: [relative_start, relative_end]
+                gt_info = [
+                    gt.start_frame / video_info['total_frames'],
+                    gt.end_frame / video_info['total_frames']
+                ]
+                if class_idx not in gts:
```

Comment:
```
gts.setdefault(class_idx, {vid:gt_info}).setdefault(vid, gt_info)
```

Reply:
`gts.setdefault(class_idx, {vid:[gt_info]})` ?

Reply:
If I use `gts.setdefault(class_idx, {vid:[gt_info]}).setdefault(vid, [gt_info]).append(gt_info)`, when the key does not exist, the value will be added twice(setdefault and append).

Reply:
can this problem be avoided?

Reply:
If we remove `append()`, then new values cannot be added.

Reply:
could you give an example?

Reply:
How about this?
```
if class_idx in gts and vid in gts[class_idx]:
    gts[class_idx][vid].append(gt_info)
else:
    gts.setdefault(class_idx, {vid:[gt_info]}).setdefault(vid, [gt_info])
```

Reply:
![image](https://user-images.githubusercontent.com/9464825/91626514-4fdd3800-e9e2-11ea-8278-93ec5093e77e.png)